kong/runloop/handler.lua

```diff
@@ -365,6 +372,190 @@ local function register_events()
+      elseif wait and wait > 0 then
+        log(NOTICE, "timeout attempting to acquire semaphore")
+      end
+
+      return false
+    end
+
+    return true
+  end
+
+
+  local function release_semaphore(semaphore)
+    semaphore:post()
+  end
+
+
+  local function log_with_worker_id(level, message)
+    if WORKER_ID then
+      log(level, message, " (worker id: ", WORKER_ID, ")")
+    else
+      log(level, message)
+    end
+  end
+
+
+  local function safe_rebuild(name, callback, version)
+    local max_attempts = 6
+    local success = false
+
+    local action = version == "init" and "initializing" or "rebuilding"
+
+    for attempt = 1, max_attempts do
+      log_with_worker_id(DEBUG,
+        fmt("%s %s. Attempt %d/%d", action, name, attempt, max_attempts))
+      local pok, ok, err = pcall(callback, version)
+      if pok and ok then
+        log_with_worker_id(DEBUG, fmt("%s %s succeeded", action, name))
+        success = true
+        break
+      end
+
+      log_with_worker_id(NOTICE, fmt("%s %s failed: %s", action, name, ok or err))
+      sleep(0.01 * attempt * attempt)
+    end
+
+    if not success then
+      return nil, fmt("%s %s failed %s times, aborting",
+                      action, name, max_attempts)
+    end
+
+    return success
+  end
+
+
+  local function rebuild_timer(premature, name, callback, version, semaphore)
+    if premature then
+      release_semaphore(semaphore)
+      return
+    end
+    local ok, err = safe_rebuild(name, callback, version)
+    if not ok then
+      log(CRIT, err)
+    end
+    release_semaphore(semaphore)
+  end
+
+
+  local function rebuild_async(name, callback, version, semaphore)
+    local ok, err = timer_at(0, rebuild_timer, name, callback, version, semaphore)
+    if not ok then
+      return nil, "could not create rebuild timer: " .. err
+    end
+
+    return true
+  end
+
+
+
+  -- @param name "router" or "plugins_iterator"
+  -- @param callback A function that will update either the router or plugins_iterator
+  -- @param version target version
+  -- @param semaphore router_semaphore or plugins_iterator_semaphore
+  -- @param wait how much time to wait when trying to acquire the semaphore.
+  -- @returns true if safe_rebuild was either successfully executed synchronously,
+  -- enqueued via async timer, or not needed (because current_version == target).
+  -- nil otherwise (safe_rebuild was neither called successfully nor enqueued).
+  -- Notice that this function might return true *even when it fails to acquire
+  -- the semaphore or invoke the async callback* - it will try to rebuild
+  -- synchronously when a wait is specified and the semaphore isn't acquired
+  -- or the timer-based call fails.
+  -- @returns error message as a second return value in case of failure
+  rebuild = function(name, callback, version, semaphore, wait)
+    local current_version, err = get_version(name)
+    if not current_version then
+      return nil, err
```

Comment:
You made the ´get_version` as a fatal error. I am not sure if that is a right call. I would in that case try to build a new router as it is done below.

Reply:
I see only one scenario where getting the version can fail: there is a problem with the cache.

The `:version` cache entry will rarely get evicted, and even if it does, a new `uuid` will be generated. So if you get a `nil` there must be a problem in the cache system (corrupted memory, that kind of thing).

If that's the case, then I think we should terminate stuff quickly. If there's a problem in the cache and there are a lot of requests arriving, we risk rebuilding the plugins iterator and the router once per request.